Fix crash when trying to send data over an already closed connection

Stacktrace from a test project:
```
[0993] DEBUG graphqlws/handler: Closing connection conn=8fab3541-9ad2-4aca-aa88-601b8d60ebb5 user=<nil>
[0993]  INFO graphqlws/subscriptions: Remove subscriptions conn=8fab3541-9ad2-4aca-aa88-601b8d60ebb5
[0993]  INFO graphqlws/subscriptions: Remove subscription conn=8fab3541-9ad2-4aca-aa88-601b8d60ebb5 subscription=1
[0993]  INFO graphqlws/subscriptions: Remove subscription conn=8fab3541-9ad2-4aca-aa88-601b8d60ebb5 subscription=2
[0993]  INFO graphqlws/subscriptions: Remove subscription conn=8fab3541-9ad2-4aca-aa88-601b8d60ebb5 subscription=8
[0993]  INFO graphqlws/subscriptions: Remove subscription conn=8fab3541-9ad2-4aca-aa88-601b8d60ebb5 subscription=4
[0993]  INFO graphqlws/subscriptions: Remove subscription conn=8fab3541-9ad2-4aca-aa88-601b8d60ebb5 subscription=10
panic: send on closed channel

goroutine 13 [running]:
github.com/functionalfoundry/graphqlws.(*connection).SendData(0xc4203095c0, 0xc420357fc8, 0x1, 0xc420b8b7a0)
	$GOPATH/go/src/github.com/functionalfoundry/graphqlws/connections.go:170 +0x134
github.com/functionalfoundry/graphqlws.NewHandler.func2.2.1(0xc4203c44d0, 0xc420b8b7a0)
	$GOPATH/go/src/github.com/functionalfoundry/graphqlws/handler.go:83 +0x4b
github.com/***/***-graphql-server/graphqlws.(*subscriptionManager).PushChange(0xc4202a0e00, 0x15609b2, 0xa, 0xc420960cc0, 0x14, 0x153e400, 0xc4204e6ff0)
	$GOPATH/go/src/github.com/***/***-graphql-server/graphqlws/subscriptions.go:123 +0xd5
main.main.func2(0x17d7360, 0xc420162680, 0x15609b2, 0xa, 0xc420960cc0, 0x14, 0x153e400, 0xc4204e6ff0)
	$GOPATH/go/src/github.com/***/***-graphql-server/main.go:104 +0x6e
github.com/***/***-graphql-server/buntdb.(*store).Set(0xc420162680, 0x15609b2, 0xa, 0xc420960cc0, 0x14, 0x153e400, 0xc4204e6ff0)
	$GOPATH/go/src/github.com/***/***-graphql-server/buntdb/store.go:107 +0x11e
github.com/***/***-graphql-server/dbchanges.NewDatabaseChangeHandler.func1(0x15609b2, 0xa, 0x153e400, 0xc4204e6ff0)
	$GOPATH/go/src/github.com/***/***-graphql-server/dbchanges/dbchanges.go:19 +0x1e5
github.com/***/***-graphql-server/mockdb.(*databaseListener).createMockData(0xc42014f570)
	$GOPATH/go/src/github.com/***/***-graphql-server/mockdb/dblistener.go:438 +0x7bb
created by github.com/***/***-graphql-server/mockdb.(*databaseListener).SetChangeHandler
	$GOPATH/go/src/github.com/***/***-graphql-server/mockdb/dblistener.go:82 +0x149
exit status 2
```

--- a/README.md
+++ b/README.md
@@ -115,7 +115,7 @@ for _, conn := range subscriptions {
 			// Errors is optional ([]error)
 			Errors: graphqlws.ErrorsFromGraphQLErrors(result.Errors),
 		}
-		subscription.SendData(subscription, &data)
+		subscription.SendData(&data)
 	}
 }
 ```

--- a/connections.go
+++ b/connections.go
@@ -4,6 +4,7 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"sync"
 	"time"
 
 	"github.com/google/uuid"
@@ -121,12 +122,14 @@ type Connection interface {
  */
 
 type connection struct {
-	id       string
-	ws       *websocket.Conn
-	config   ConnectionConfig
-	logger   *log.Entry
-	outgoing chan OperationMessage
-	user     interface{}
+	id         string
+	ws         *websocket.Conn
+	config     ConnectionConfig
+	logger     *log.Entry
+	outgoing   chan OperationMessage
+	user       interface{}
+	closeMutex *sync.Mutex
+	closed     bool
 }
 
 func operationMessageForType(messageType string) OperationMessage {
@@ -144,6 +147,8 @@ func NewConnection(ws *websocket.Conn, config ConnectionConfig) Connection {
 	conn.ws = ws
 	conn.config = config
 	conn.logger = NewLogger("connection/" + conn.id)
+	conn.closed = false
+	conn.closeMutex = &sync.Mutex{}
 
 	conn.outgoing = make(chan OperationMessage)
 
@@ -167,25 +172,43 @@ func (conn *connection) SendData(opID string, data *DataMessagePayload) {
 	msg := operationMessageForType(gqlData)
 	msg.ID = opID
 	msg.Payload = data
-	conn.outgoing <- msg
+	conn.closeMutex.Lock()
+	if !conn.closed {
+		conn.outgoing <- msg
+	}
+	conn.closeMutex.Unlock()
 }
 
 func (conn *connection) SendError(err error) {
 	msg := operationMessageForType(gqlError)
 	msg.Payload = err.Error()
-	conn.outgoing <- msg
+	conn.closeMutex.Lock()
+	if !conn.closed {
+		conn.outgoing <- msg
+	}
+	conn.closeMutex.Unlock()
 }
 
 func (conn *connection) sendOperationErrors(opID string, errs []error) {
+	if conn.closed {
+		return
+	}
 	msg := operationMessageForType(gqlError)
 	msg.ID = opID
 	msg.Payload = errs
-	conn.outgoing <- msg
+	conn.closeMutex.Lock()
+	if !conn.closed {
+		conn.outgoing <- msg
+	}
+	conn.closeMutex.Unlock()
 }
 
 func (conn *connection) close() {
 	// Close the write loop by closing the outgoing messages channels
+	conn.closeMutex.Lock()
+	conn.closed = true
 	close(conn.outgoing)
+	conn.closeMutex.Unlock()
 
 	// Notify event handlers
 	if conn.config.EventHandlers.Close != nil {

--- a/handler.go
+++ b/handler.go
@@ -79,7 +79,7 @@ func NewHandler(config HandlerConfig) http.Handler {
 							Variables:     data.Variables,
 							OperationName: data.OperationName,
 							Connection:    conn,
-							SendData: func(subscription *Subscription, data *DataMessagePayload) {
+							SendData: func(data *DataMessagePayload) {
 								conn.SendData(opID, data)
 							},
 						})

--- a/subscriptions.go
+++ b/subscriptions.go
@@ -21,7 +21,7 @@ func ErrorsFromGraphQLErrors(errors []gqlerrors.FormattedError) []error {
 
 // SubscriptionSendDataFunc is a function that sends updated data
 // for a specific subscription to the corresponding subscriber.
-type SubscriptionSendDataFunc func(*Subscription, *DataMessagePayload)
+type SubscriptionSendDataFunc func(*DataMessagePayload)
 
 // Subscription holds all information about a GraphQL subscription
 // made by a client, including a function to send data back to the

--- a/subscriptions_test.go
+++ b/subscriptions_test.go
@@ -125,7 +125,7 @@ func TestSubscriptions_AddingValidSubscriptionsWorks(t *testing.T) {
 		ID:         "1",
 		Connection: &conn,
 		Query:      "subscription { users }",
-		SendData: func(s *graphqlws.Subscription, msg *graphqlws.DataMessagePayload) {
+		SendData: func(msg *graphqlws.DataMessagePayload) {
 			// Do nothing
 		},
 	}
@@ -149,7 +149,7 @@ func TestSubscriptions_AddingValidSubscriptionsWorks(t *testing.T) {
 		ID:         "2",
 		Connection: &conn,
 		Query:      "subscription { users }",
-		SendData: func(s *graphqlws.Subscription, msg *graphqlws.DataMessagePayload) {
+		SendData: func(msg *graphqlws.DataMessagePayload) {
 			// Do nothing
 		},
 	}
@@ -189,7 +189,7 @@ func TestSubscriptions_AddingSubscriptionsTwiceFails(t *testing.T) {
 		ID:         "1",
 		Connection: &conn,
 		Query:      "subscription { users }",
-		SendData: func(s *graphqlws.Subscription, msg *graphqlws.DataMessagePayload) {
+		SendData: func(msg *graphqlws.DataMessagePayload) {
 			// Do nothing
 		},
 	}
@@ -233,7 +233,7 @@ func TestSubscriptions_RemovingSubscriptionsWorks(t *testing.T) {
 		ID:         "1",
 		Connection: &conn,
 		Query:      "subscription { users }",
-		SendData: func(s *graphqlws.Subscription, msg *graphqlws.DataMessagePayload) {
+		SendData: func(msg *graphqlws.DataMessagePayload) {
 			// Do nothing
 		},
 	}
@@ -242,7 +242,7 @@ func TestSubscriptions_RemovingSubscriptionsWorks(t *testing.T) {
 		ID:         "2",
 		Connection: &conn,
 		Query:      "subscription { users }",
-		SendData: func(s *graphqlws.Subscription, msg *graphqlws.DataMessagePayload) {
+		SendData: func(msg *graphqlws.DataMessagePayload) {
 			// Do nothing
 		},
 	}
@@ -285,7 +285,7 @@ func TestSubscriptions_RemovingSubscriptionsOfAConnectionWorks(t *testing.T) {
 		ID:         "1",
 		Connection: &conn1,
 		Query:      "subscription { users }",
-		SendData: func(s *graphqlws.Subscription, msg *graphqlws.DataMessagePayload) {
+		SendData: func(msg *graphqlws.DataMessagePayload) {
 			// Do nothing
 		},
 	}
@@ -294,7 +294,7 @@ func TestSubscriptions_RemovingSubscriptionsOfAConnectionWorks(t *testing.T) {
 		ID:         "2",
 		Connection: &conn1,
 		Query:      "subscription { users }",
-		SendData: func(s *graphqlws.Subscription, msg *graphqlws.DataMessagePayload) {
+		SendData: func(msg *graphqlws.DataMessagePayload) {
 			// Do nothing
 		},
 	}
@@ -303,7 +303,7 @@ func TestSubscriptions_RemovingSubscriptionsOfAConnectionWorks(t *testing.T) {
 		ID:         "1",
 		Connection: &conn2,
 		Query:      "subscription { users }",
-		SendData: func(s *graphqlws.Subscription, msg *graphqlws.DataMessagePayload) {
+		SendData: func(msg *graphqlws.DataMessagePayload) {
 			// Do nothing
 		},
 	}
@@ -312,7 +312,7 @@ func TestSubscriptions_RemovingSubscriptionsOfAConnectionWorks(t *testing.T) {
 		ID:         "2",
 		Connection: &conn2,
 		Query:      "subscription { users }",
-		SendData: func(s *graphqlws.Subscription, msg *graphqlws.DataMessagePayload) {
+		SendData: func(msg *graphqlws.DataMessagePayload) {
 			// Do nothing
 		},
 	}

